src/Materials/Textures/Packer/packer.ts

```diff
@@ -0,0 +1,761 @@
+    private _disposeList: (Texture | DynamicTexture)[];
+
+    /** The padding value from Math.ceil(frameSize * paddingRatio) */
+    private _paddingValue: number;
+
+    /**
+    * Initializes a texture package series from an array of meshes or a single mesh.
+    * @param name The name of the package
+    * @param meshes The target meshes to compose the package from
+    * @param options The arguments that texture packer should follow while building.
+    * @param scene The scene which the textures are scoped to.
+    * @returns TexturePacker
+    */
+    constructor(name: string, meshes: AbstractMesh[], options: ITexturePackerOptions, scene: Scene) {
+
+        this.name = name;
+        this.meshes = meshes;
+        this.scene = scene;
+
+        /**
+        * Run through the options and set what ever defaults are needed that where not declared.
+        */
+        this.options = options;
+        this.options.map = this.options.map || [
+                'ambientTexture',
+                'bumpTexture',
+                'diffuseTexture',
+                'emissiveTexture',
+                'lightmapTexture',
+                'opacityTexture',
+                'reflectionTexture',
+                'refractionTexture',
+                'specularTexture'
+            ];
+
+        this.options.uvsIn = this.options.uvsIn || VertexBuffer.UVKind;
+        this.options.uvsOut = this.options.uvsOut || VertexBuffer.UVKind;
+        this.options.layout = this.options.layout || TexturePacker.LAYOUT_STRIP;
+
+        if (this.options.layout === TexturePacker.LAYOUT_COLNUM) {
+            this.options.colnum = this.options.colnum || 8;
+        }
+
+        this.options.updateInputMeshes = this.options.updateInputMeshes || true;
+        this.options.disposeSources = this.options.disposeSources || true;
+        this._disposeList = [];
+
+        this.options.fillBlanks = this.options.fillBlanks || true;
+
+        if (this.options.fillBlanks === true) {
+            this.options.customFillColor = this.options.customFillColor || 'black';
+        }
+
+        this.options.frameSize = this.options.frameSize || 256;
+        this.options.paddingRatio = this.options.paddingRatio || 0.0115;
+
+        this._paddingValue = Math.ceil(this.options.frameSize * this.options.paddingRatio);
+
+        //Make it an even padding Number.
+        if (this._paddingValue % 2 !== 0) {
+            this._paddingValue++;
+        }
+
+        this.options.paddingMode = this.options.paddingMode || TexturePacker.SUBUV_WRAP;
+
+        if (this.options.paddingMode === TexturePacker.SUBUV_COLOR) {
+            this.options.paddingColor = this.options.paddingColor || new Color4(0, 0, 0, 1.0);
+        }
+
+        this.sets = {};
+        this.frames = [];
+
+        /**
+        * Create the promise and then run through the materials on the meshes.
+        */
+        this.promise = new Promise ((resolve, reject) => {

```

Comment:
I wonder if instead of having a then function which is basically a way to give resolve/error to this promise we should not have a texturePacker.processAsync() which will return a promise. So the user can use then or catch or anything related to a promise like using Promise.all if he/she has multiple packers

Reply:
sure?  If you think that is better I kinda thought the promise let it resolve in async like a doing a setTimeout hack.

Reply:
I still want the promise but it should not be set in the constructor and "awaited" by a fake then function. It should instead be created and returned by a processAsync function . In a nutshell:
- remove the then function
- add a processAsync function and move the promise there
- return the promise

Reply:
Ok I'll have to look up how to do that, should be EZPZ though!

So the user is going to have to say when to start packing basically?

Reply:
Yes and I usually prefer that. Creating an object should not start any heavy duty process:)

Reply:
I will have that done this morning.

Reply:
I will have that done this morning.

Reply:
So basically the user will create the TexturePacker, but will have to manually call `pack.processAsync()` which is where most the stuff I am doing in the constructor will get moved?

That will also mean then I get rid of the Loader class like mentioned and then just have them access that class directly which I can then add an argument to have that start its async Compilation as well.

Am I understanding you correctly?

Reply:
totally

Reply:
hmm I've never used an await before, let me do some reading on this.

Reply:
no need for await (And technically we cannot use them as this is not supported on IE and we have no polyfill for it)

Reply:
Ok cool! I was kinda in full research mode for a second and then was just about to ask about compatibility.